Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc

```diff
@@ -10434,4 +10434,243 @@ int4 RuleLzcountShiftBool::applyOp(PcodeOp *op,Funcdata &data)
+
+{
+  if (!data.hasTypeRecoveryStarted()) return 0;
+
+  list<PcodeOp *>::const_iterator iter;
+  PcodeOp *op;
+
+  const BlockGraph &basicblocks( data.getBasicBlocks() );
+  for(int4 j=0;j<basicblocks.getSize();++j) {
+    BlockBasic *bb = (BlockBasic *)basicblocks.getBlock(j);
+    for(iter=bb->beginOp();iter!=bb->endOp();++iter) {
+      op = *iter;
+      OpCode code = op->code();
+      if (code != CPUI_STORE
+	&& code != CPUI_INT_AND
+	&& code != CPUI_INT_OR
+	&& code != CPUI_INT_EQUAL         // for enums that are downcast to smaller types
+	&& code != CPUI_INT_NOTEQUAL) {   // for enums that are downcast to smaller types
+	continue;
+      }
+
+      Varnode* constant = op->getIn(1);
+      if (op->code() == CPUI_STORE) {
+	constant = op->getIn(2);
+      }
+      if (!constant->isConstant() || constant->getType()->isEnumType() || constant->isSpacebase() || constant->isTypeLock()) {
+	continue;
+      }
+      PcodeOp* def = op;
+      if (op->code() != CPUI_STORE) {
+	def = op->getIn(0)->getDef();
+	if (!def) {
+	  Varnode* opOut = op->getOut();
+	  if (opOut) {
+	    def = opOut->loneDescend();
+	  }
+	}
+      }
+      if (!def) continue;
+
+      struct VarnodeToScan {
+	Varnode* varnode = nullptr;
+	int operationSize = 0;
+	bool subpieceUpper4Bytes = false;
+	int slot = 0;
+      };
+
+      vector<VarnodeToScan> varnodes;
+
+      code = def->code();
+
+      if (code == CPUI_LOAD) {
+	varnodes.emplace_back();
+	VarnodeToScan& newVnToScan = varnodes.back();
+	newVnToScan.varnode = def->getIn(1);
+	newVnToScan.operationSize = def->getOut()->getSize();
+	newVnToScan.slot = 1;
+
+      } else if (code == CPUI_SUBPIECE) {
+	varnodes.emplace_back();
+	VarnodeToScan& newVnToScan = varnodes.back();
+	newVnToScan.varnode = def->getIn(0);
+	newVnToScan.slot = 0;
+	Varnode* subpiece = def->getIn(1);
+	uintb subpieceOffset = subpiece->getOffset();
+	if (subpieceOffset != 0 && subpieceOffset != 4) {
+	  continue;
+	}
+	newVnToScan.subpieceUpper4Bytes = (subpieceOffset == 4);
+	newVnToScan.operationSize = subpiece->getSize();
+
+      } else if (code == CPUI_STORE) {
+	varnodes.emplace_back();
+	VarnodeToScan& newVnToScan = varnodes.back();
+	newVnToScan.varnode = def->getIn(1);
+	newVnToScan.slot = 1;
+	newVnToScan.operationSize = def->getIn(2)->getSize();
+
+      } else if (code == CPUI_INT_AND || code == CPUI_INT_OR) {
+	int operationSize = def->getOut()->getSize();
+	for (int i = 0; i < def->numInput(); ++i) {
+	  varnodes.emplace_back();
+	  VarnodeToScan& newVnToScan = varnodes.back();
+	  newVnToScan.varnode = def->getIn(i);
+	  newVnToScan.slot = i;
+	  newVnToScan.operationSize = operationSize;
+	}
+
+      } else if (code == CPUI_MULTIEQUAL
+	  || code == CPUI_CALL
+	  || code == CPUI_CALLIND) {
+	varnodes.emplace_back();
+	VarnodeToScan& newVnToScan = varnodes.back();
+	newVnToScan.varnode = def->getOut();
+	newVnToScan.slot = -1;
+	newVnToScan.operationSize = def->getOut()->getSize();
```

Comment:
It can crash here if `def->getOut() == 0`:
```
Program received signal SIGSEGV, Segmentation fault.
ghidra::Varnode::getSize (this=0x0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/include/ghidra/varnode.hh:185
185       int4 getSize(void) const { return size; } ///< Get the number of bytes this Varnode stores
(gdb) bt full
#0  ghidra::Varnode::getSize (this=0x0) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/include/ghidra/varnode.hh:185
No locals.
#1  0x000055aaa79e18a9 in ghidra::ActionPropagateEnums::apply (this=0x55aaca1268c0, data=...) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc:11132
        newVnToScan = @0x55aaca129b80: {varnode = 0x0, operationSize = 0, subpieceUpper4Bytes = false, slot = -1}
        code = ghidra::CPUI_CALL
        constant = 0x55aaca136520
        def = 0x55aaca129ff0
        varnodes = std::vector of length 1, capacity 1 = {{varnode = 0x0, operationSize = 0, subpieceUpper4Bytes = false, slot = -1}}
        bb = 0x55aaca12e040
        j = 2
        iter = 0x55aaca139f90
        op = 0x55aaca139f90
        basicblocks = @0x55aaca157a28: {<ghidra::FlowBlock> = {_vptr$FlowBlock = 0x55aaa79f8b28 <vtable for ghidra::BlockGraph+16>, flags = 0, parent = 0x0, immed_dom = 0x0, 
            copymap = 0x6574756269, index = 0, visitcount = 0, numdesc = 8, intothis = std::vector of length 0, capacity 0, outofthis = std::vector of length 0, capacity 0}, 
          list = std::vector of length 4, capacity 4 = {0x55aaca12d9c0, 0x55aaca12dde0, 0x55aaca12e040, 0x55aaca12e220}}
#2  0x000055aaa76b08ad in ghidra::Action::perform (this=0x55aaca1268c0, data=...) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/action.cc:319
        res = 32766
#3  0x000055aaa76b153e in ghidra::ActionGroup::apply (this=0x55aaca116bf0, data=...) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/action.cc:512
        res = 0
#4  0x000055aaa76b178c in ghidra::ActionRestartGroup::apply (this=0x55aaca116bf0, data=...) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/action.cc:558
        iter = 0x7ffe423356f0
        res = 32766
#5  0x000055aaa76b08ad in ghidra::Action::perform (this=0x55aaca116bf0, data=...) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/action.cc:319
        res = 32766
#6  0x000055aaa77c9098 in ghidra::DecompileAt::rawAction (this=0x55aaca0fbe00) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/ghidra_process.cc:317
        fd = 0x55aaca157740
#7  0x000055aaa77c81ba in ghidra::GhidraCommand::doit (this=0x55aaca0fbe00) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/ghidra_process.cc:144
        type = 3
        err = <error reading variable: Cannot access memory at address 0x0>
        err = @0x7fda79360dec: {<ghidra::LowlevelError> = {explain = <incomplete type>}, type = <incomplete type>}
        err = @0x7ffe42335b40: {<ghidra::LowlevelError> = {explain = <incomplete type>}, <No data fields>}
        err = @0x55aaa76b8ba1: {explain = <incomplete type>}
#8  0x000055aaa77ca0e1 in ghidra::GhidraCapability::readCommand (sin=..., out=...) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/ghidra_process.cc:491
        function = <incomplete type>
        type = 2
        iter = {first = <incomplete type>, second = 0x55aaca0fbe00}
#9  0x000055aaa77caacf in main (argc=1, argv=0x7ffe42335df8) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/ghidra_process.cc:533
        status = 0
(gdb) frame 1
#1  0x000055aaa79e18a9 in ghidra::ActionPropagateEnums::apply (this=0x55aaca1268c0, data=...) at /tmp/ghidra/Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc:11132
11132           newVnToScan.operationSize = def->getOut()->getSize();
(gdb) print def->getOut()
$1 = (ghidra::Varnode *) 0x0
```

Reply:
@jobermayr Thank you! I've made a commit that adds checks for 0 on this.